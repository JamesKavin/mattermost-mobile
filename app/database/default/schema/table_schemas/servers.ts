// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {tableSchema} from '@nozbe/watermelondb';

import {MM_TABLES} from '@constants/database';

const {SERVERS} = MM_TABLES.DEFAULT;

export default tableSchema({
    name: SERVERS,
    columns: [
        {name: 'db_path', type: 'string'},
        {name: 'display_name', type: 'string'},
        {name: 'mention_count', type: 'number'},
        {name: 'unread_count', type: 'number'},
        {name: 'url', type: 'string', isIndexed: true},
    ],
});
