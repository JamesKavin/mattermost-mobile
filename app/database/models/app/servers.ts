// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Model} from '@nozbe/watermelondb';
import {field} from '@nozbe/watermelondb/decorators';

import {MM_TABLES} from '@constants/database';

const {SERVERS} = MM_TABLES.APP;

/**
 * The Server model will help us to identify the various servers a user will log in; in the context of
 * multi-server support system.  The db_path field will hold the App-Groups file-path
 */
export default class Servers extends Model {
    /** table (name) : servers */
    static table = SERVERS;

    /** db_path : The file path where the database is stored */
    @field('db_path') dbPath!: string;

    /** display_name : The server display name */
    @field('display_name') displayName!: string;

    /** mention_count : The number of mention on this server */
    @field('mention_count') mentionCount!: number;

    /** unread_count : The number of unread messages on this server */
    @field('unread_count') unreadCount!: number;

    /** url : The online address for the Mattermost server */
    @field('url') url!: string;

    /** last_active_at: The last time this server was active */
    @field('last_active_at') lastActiveAt!: number;

    /** is_secured: Determines if the protocol used for this server url is HTTP or HTTPS */
    @field('is_secured') isSecured!: boolean;
}
