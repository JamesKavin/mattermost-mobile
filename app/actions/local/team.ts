// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DatabaseManager from '@database/manager';
import {prepareDeleteTeam, queryMyTeamById, removeTeamFromTeamHistory} from '@queries/servers/team';

import type TeamModel from '@typings/database/models/servers/team';

export const removeUserFromTeam = async (serverUrl: string, teamId: string) => {
    const serverDatabase = DatabaseManager.serverDatabases[serverUrl];
    if (!serverDatabase) {
        return;
    }

    const {operator, database} = serverDatabase;

    const myTeam = await queryMyTeamById(database, teamId);
    if (myTeam) {
        const team = await myTeam.team.fetch() as TeamModel;
        const models = await prepareDeleteTeam(team);
        const system = await removeTeamFromTeamHistory(operator, team.id, true);
        if (system) {
            models.push(...system);
        }
        if (models.length) {
            try {
                await operator.batchRecords(models);
            } catch {
                // eslint-disable-next-line no-console
                console.log('FAILED TO BATCH CHANGES FOR REMOVE USER FROM TEAM');
            }
        }
    }
};
