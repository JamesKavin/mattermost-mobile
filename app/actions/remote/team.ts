// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Model} from '@nozbe/watermelondb';
import {DeviceEventEmitter} from 'react-native';

import {removeUserFromTeam as localRemoveUserFromTeam} from '@actions/local/team';
import {Client} from '@client/rest';
import {PER_PAGE_DEFAULT} from '@client/rest/constants';
import {Events} from '@constants';
import DatabaseManager from '@database/manager';
import NetworkManager from '@managers/network_manager';
import {getActiveServerUrl} from '@queries/app/servers';
import {prepareCategoriesAndCategoriesChannels} from '@queries/servers/categories';
import {prepareMyChannelsForTeam, getDefaultChannelForTeam} from '@queries/servers/channel';
import {prepareCommonSystemValues, getCurrentTeamId, getCurrentUserId} from '@queries/servers/system';
import {addTeamToTeamHistory, prepareDeleteTeam, prepareMyTeams, getNthLastChannelFromTeam, queryTeamsById, getLastTeam, getTeamById, removeTeamFromTeamHistory, queryMyTeams} from '@queries/servers/team';
import {dismissAllModals, popToRoot} from '@screens/navigation';
import EphemeralStore from '@store/ephemeral_store';
import {setTeamLoading} from '@store/team_load_store';
import {isTablet} from '@utils/helpers';
import {logDebug} from '@utils/log';

import {fetchMyChannelsForTeam, switchToChannelById} from './channel';
import {fetchGroupsForTeamIfConstrained} from './groups';
import {fetchPostsForChannel, fetchPostsForUnreadChannels} from './post';
import {fetchRolesIfNeeded} from './role';
import {forceLogoutIfNecessary} from './session';

import type ClientError from '@client/rest/error';

export type MyTeamsRequest = {
    teams?: Team[];
    memberships?: TeamMembership[];
    error?: unknown;
}

export async function addCurrentUserToTeam(serverUrl: string, teamId: string, fetchOnly = false) {
    let database;
    try {
        database = DatabaseManager.getServerDatabaseAndOperator(serverUrl).database;
    } catch (error) {
        return {error};
    }

    const currentUserId = await getCurrentUserId(database);

    if (!currentUserId) {
        return {error: 'no current user'};
    }
    return addUserToTeam(serverUrl, teamId, currentUserId, fetchOnly);
}

export async function addUserToTeam(serverUrl: string, teamId: string, userId: string, fetchOnly = false) {
    let client;
    try {
        client = NetworkManager.getClient(serverUrl);
    } catch (error) {
        return {error};
    }

    let loadEventSent = false;
    try {
        EphemeralStore.startAddingToTeam(teamId);
        const team = await client.getTeam(teamId);
        const member = await client.addToTeam(teamId, userId);

        if (!fetchOnly) {
            setTeamLoading(serverUrl, true);
            loadEventSent = true;

            fetchRolesIfNeeded(serverUrl, member.roles.split(' '));
            const {channels, memberships: channelMembers, categories} = await fetchMyChannelsForTeam(serverUrl, teamId, false, 0, true);
            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
            if (operator) {
                const myTeams: MyTeam[] = [{
                    id: member.team_id,
                    roles: member.roles,
                }];

                const models: Model[] = (await Promise.all([
                    operator.handleTeam({teams: [team], prepareRecordsOnly: true}),
                    operator.handleMyTeam({myTeams, prepareRecordsOnly: true}),
                    operator.handleTeamMemberships({teamMemberships: [member], prepareRecordsOnly: true}),
                    ...await prepareMyChannelsForTeam(operator, teamId, channels || [], channelMembers || []),
                    prepareCategoriesAndCategoriesChannels(operator, categories || [], true),
                ])).flat();

                await operator.batchRecords(models);
                setTeamLoading(serverUrl, false);
                loadEventSent = false;

                if (await isTablet()) {
                    const channel = await getDefaultChannelForTeam(operator.database, teamId);
                    if (channel) {
                        fetchPostsForChannel(serverUrl, channel.id);
                    }
                }
            } else {
                setTeamLoading(serverUrl, false);
                loadEventSent = false;
            }
        }
        EphemeralStore.finishAddingToTeam(teamId);
        updateCanJoinTeams(serverUrl);
        return {member};
    } catch (error) {
        if (loadEventSent) {
            setTeamLoading(serverUrl, false);
        }
        EphemeralStore.finishAddingToTeam(teamId);
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
}

export async function fetchMyTeams(serverUrl: string, fetchOnly = false): Promise<MyTeamsRequest> {
    let client;
    try {
        client = NetworkManager.getClient(serverUrl);
    } catch (error) {
        return {error};
    }

    try {
        const [teams, memberships]: [Team[], TeamMembership[]] = await Promise.all([
            client.getMyTeams(),
            client.getMyTeamMembers(),
        ]);

        if (!fetchOnly) {
            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
            const modelPromises: Array<Promise<Model[]>> = [];
            if (operator) {
                const removeTeamIds = new Set(memberships.filter((m) => m.delete_at > 0).map((m) => m.team_id));
                const remainingTeams = teams.filter((t) => !removeTeamIds.has(t.id));
                const prepare = prepareMyTeams(operator, remainingTeams, memberships);
                if (prepare) {
                    modelPromises.push(...prepare);
                }

                if (removeTeamIds.size) {
                    // Immediately delete myTeams so that the UI renders only teams the user is a member of.
                    const removeTeams = await queryTeamsById(operator.database, Array.from(removeTeamIds)).fetch();
                    removeTeams.forEach((team) => {
                        modelPromises.push(prepareDeleteTeam(team));
                    });
                }

                if (modelPromises.length) {
                    const models = await Promise.all(modelPromises);
                    const flattenedModels = models.flat();
                    if (flattenedModels.length > 0) {
                        await operator.batchRecords(flattenedModels);
                    }
                }
            }
        }

        return {teams, memberships};
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
}

export async function fetchMyTeam(serverUrl: string, teamId: string, fetchOnly = false): Promise<MyTeamsRequest> {
    let client;
    try {
        client = NetworkManager.getClient(serverUrl);
    } catch (error) {
        return {error};
    }

    try {
        const [team, membership] = await Promise.all([
            client.getTeam(teamId),
            client.getTeamMember(teamId, 'me'),
        ]);
        if (!fetchOnly) {
            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
            if (operator) {
                const modelPromises = prepareMyTeams(operator, [team], [membership]);
                if (modelPromises.length) {
                    const models = await Promise.all(modelPromises);
                    const flattenedModels = models.flat();
                    if (flattenedModels?.length > 0) {
                        await operator.batchRecords(flattenedModels);
                    }
                }
            }
        }

        return {teams: [team], memberships: [membership]};
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
}

export const fetchAllTeams = async (serverUrl: string, page = 0, perPage = PER_PAGE_DEFAULT): Promise<{teams?: Team[]; error?: any}> => {
    try {
        const client = NetworkManager.getClient(serverUrl);
        const teams = await client.getTeams(page, perPage);
        return {teams};
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
};

const recCanJoinTeams = async (client: Client, myTeamsIds: Set<string>, page: number): Promise<boolean> => {
    const fetchedTeams = await client.getTeams(page, PER_PAGE_DEFAULT);
    if (fetchedTeams.find((t) => !myTeamsIds.has(t.id) && t.delete_at === 0)) {
        return true;
    }

    if (fetchedTeams.length === PER_PAGE_DEFAULT) {
        return recCanJoinTeams(client, myTeamsIds, page + 1);
    }

    return false;
};

const LOAD_MORE_THRESHOLD = 10;
export async function fetchTeamsForComponent(
    serverUrl: string,
    page: number,
    joinedIds?: Set<string>,
    alreadyLoaded: Team[] = [],
): Promise<{teams: Team[]; hasMore: boolean; page: number}> {
    let hasMore = true;
    const {teams, error} = await fetchAllTeams(serverUrl, page, PER_PAGE_DEFAULT);
    if (error || !teams || teams.length < PER_PAGE_DEFAULT) {
        hasMore = false;
    }

    if (error) {
        return {teams: alreadyLoaded, hasMore, page};
    }

    if (teams?.length) {
        const notJoinedTeams = joinedIds ? teams.filter((t) => !joinedIds.has(t.id)) : teams;
        alreadyLoaded.push(...notJoinedTeams);

        if (teams.length < PER_PAGE_DEFAULT) {
            hasMore = false;
        }

        if (
            hasMore &&
            (alreadyLoaded.length > LOAD_MORE_THRESHOLD)
        ) {
            return fetchTeamsForComponent(serverUrl, page + 1, joinedIds, alreadyLoaded);
        }

        return {teams: alreadyLoaded, hasMore, page: page + 1};
    }

    return {teams: alreadyLoaded, hasMore: false, page};
}

export const updateCanJoinTeams = async (serverUrl: string) => {
    try {
        const client = NetworkManager.getClient(serverUrl);
        const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);

        const myTeams = await queryMyTeams(database).fetch();
        const myTeamsIds = new Set(myTeams.map((m) => m.id));

        const canJoin = await recCanJoinTeams(client, myTeamsIds, 0);

        EphemeralStore.setCanJoinOtherTeams(serverUrl, canJoin);
        return {};
    } catch (error) {
        EphemeralStore.setCanJoinOtherTeams(serverUrl, false);
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
};

export const fetchTeamsChannelsAndUnreadPosts = async (serverUrl: string, since: number, teams: Team[], memberships: TeamMembership[], excludeTeamId?: string) => {
    const database = DatabaseManager.serverDatabases[serverUrl]?.database;
    if (!database) {
        return {error: `${serverUrl} database not found`};
    }

    const membershipSet = new Set(memberships.map((m) => m.team_id));
    const myTeams = teams.filter((t) => membershipSet.has(t.id) && t.id !== excludeTeamId);

    for await (const team of myTeams) {
        const {channels, memberships: members} = await fetchMyChannelsForTeam(serverUrl, team.id, true, since, false, true);

        if (channels?.length && members?.length) {
            fetchPostsForUnreadChannels(serverUrl, channels, members);
        }
    }

    return {error: undefined};
};

export async function fetchTeamByName(serverUrl: string, teamName: string, fetchOnly = false) {
    let client;
    try {
        client = NetworkManager.getClient(serverUrl);
    } catch (error) {
        return {error};
    }

    try {
        const team = await client.getTeamByName(teamName);

        if (!fetchOnly) {
            const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
            if (operator) {
                const models = await operator.handleTeam({teams: [team], prepareRecordsOnly: true});
                await operator.batchRecords(models);
            }
        }

        return {team};
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
}

export const removeCurrentUserFromTeam = async (serverUrl: string, teamId: string, fetchOnly = false) => {
    try {
        const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const userId = await getCurrentUserId(database);
        return removeUserFromTeam(serverUrl, teamId, userId, fetchOnly);
    } catch (error) {
        return {error};
    }
};

export const removeUserFromTeam = async (serverUrl: string, teamId: string, userId: string, fetchOnly = false) => {
    let client;
    try {
        client = NetworkManager.getClient(serverUrl);
    } catch (error) {
        return {error};
    }

    try {
        await client.removeFromTeam(teamId, userId);

        if (!fetchOnly) {
            localRemoveUserFromTeam(serverUrl, teamId);
            updateCanJoinTeams(serverUrl);
        }

        return {error: undefined};
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error as ClientError);
        return {error};
    }
};

export async function handleTeamChange(serverUrl: string, teamId: string) {
    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
    if (!operator) {
        return;
    }

    const {database} = operator;
    const currentTeamId = await getCurrentTeamId(database);

    if (currentTeamId === teamId) {
        return;
    }

    let channelId = '';
    DeviceEventEmitter.emit(Events.TEAM_SWITCH, true);
    if (await isTablet()) {
        channelId = await getNthLastChannelFromTeam(database, teamId);
        if (channelId) {
            await switchToChannelById(serverUrl, channelId, teamId);
            DeviceEventEmitter.emit(Events.TEAM_SWITCH, false);
            return;
        }
    }

    const models = [];
    const system = await prepareCommonSystemValues(operator, {currentChannelId: channelId, currentTeamId: teamId, lastUnreadChannelId: ''});
    if (system?.length) {
        models.push(...system);
    }
    const history = await addTeamToTeamHistory(operator, teamId, true);
    if (history.length) {
        models.push(...history);
    }

    if (models.length) {
        await operator.batchRecords(models);
    }
    DeviceEventEmitter.emit(Events.TEAM_SWITCH, false);

    // Fetch Groups + GroupTeams
    fetchGroupsForTeamIfConstrained(serverUrl, teamId);
}

export async function handleKickFromTeam(serverUrl: string, teamId: string) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const currentTeamId = await getCurrentTeamId(database);
        if (currentTeamId !== teamId) {
            return;
        }

        const currentServer = await getActiveServerUrl();
        if (currentServer === serverUrl) {
            const team = await getTeamById(database, teamId);
            DeviceEventEmitter.emit(Events.LEAVE_TEAM, team?.displayName);
            await dismissAllModals();
            await popToRoot();
        }

        await removeTeamFromTeamHistory(operator, teamId);
        const teamToJumpTo = await getLastTeam(database, teamId);
        if (teamToJumpTo) {
            await handleTeamChange(serverUrl, teamToJumpTo);
        }

        // Resetting to team select handled by the home screen
    } catch (error) {
        logDebug('Failed to kick user from team', error);
    }
}
