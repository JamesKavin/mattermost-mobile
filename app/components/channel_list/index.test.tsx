// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Database from '@nozbe/watermelondb/Database';
import React from 'react';
import {SafeAreaProvider} from 'react-native-safe-area-context';

import {SYSTEM_IDENTIFIERS} from '@constants/database';
import ServerDataOperator from '@database/operator/server_data_operator';
import {getTeamById} from '@queries/servers/team';
import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import ChannelsList from './';

describe('components/channel_list', () => {
    let database: Database;
    let operator: ServerDataOperator;
    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
        operator = server.operator;

        const team = await getTeamById(database, TestHelper.basicTeam!.id);
        await database.write(async () => {
            await team?.update(() => {
                team.displayName = 'Test Team!';
            });
        });
    });

    it('should render', () => {
        const wrapper = renderWithEverything(
            <SafeAreaProvider>
                <ChannelsList
                    isTablet={false}
                    teamsCount={1}
                    currentTeamId={TestHelper.basicTeam!.id}
                    channelsCount={1}
                />
            </SafeAreaProvider>,
            {database},
        );
        expect(wrapper.toJSON()).toBeTruthy();
    });

    it('should render team error', async () => {
        await operator.handleSystem({
            systems: [{id: SYSTEM_IDENTIFIERS.CURRENT_TEAM_ID, value: ''}],
            prepareRecordsOnly: false,
        });

        const wrapper = renderWithEverything(
            <SafeAreaProvider>
                <ChannelsList
                    isTablet={false}
                    teamsCount={0}
                    currentTeamId='TestHelper.basicTeam!.id'
                    channelsCount={1}
                />
            </SafeAreaProvider>,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

        await operator.handleSystem({
            systems: [{id: SYSTEM_IDENTIFIERS.CURRENT_TEAM_ID, value: TestHelper.basicTeam!.id}],
            prepareRecordsOnly: false,
        });
    });

    it('should render channels error', () => {
        const wrapper = renderWithEverything(
            <SafeAreaProvider>
                <ChannelsList
                    isTablet={false}
                    teamsCount={1}
                    currentTeamId={TestHelper.basicTeam!.id}
                    channelsCount={0}
                />
            </SafeAreaProvider>,
            {database},
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
