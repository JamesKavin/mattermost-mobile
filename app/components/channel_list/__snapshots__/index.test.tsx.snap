// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/channel_list should render channels error 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "maxWidth": "100%",
      },
    }
  }
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#1e325c",
      "flex": 1,
      "maxWidth": "100%",
      "paddingLeft": 18,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "marginLeft": 0,
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "marginLeft": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "Metropolis-SemiBold",
                "fontSize": 28,
                "fontWeight": "600",
                "lineHeight": 36,
              }
            }
            testID="channel_list_header.team_display_name"
          >
            Test Team!
          </Text>
          <View
            style={
              Object {
                "marginLeft": 4,
              }
            }
          >
            <Icon
              name="chevron-down"
              style={
                Object {
                  "color": "rgba(255,255,255,0.8)",
                  "fontSize": 24,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.08)",
            "borderRadius": 14,
            "height": 28,
            "justifyContent": "center",
            "opacity": 1,
            "width": 28,
          }
        }
      >
        <Icon
          name="plus"
          style={
            Object {
              "color": "rgba(255,255,255,0.8)",
              "fontSize": 18,
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "rgba(255,255,255,0.64)",
          "fontFamily": "OpenSans-SemiBold",
          "fontSize": 11,
          "fontWeight": "600",
          "lineHeight": 16,
        }
      }
      testID="channel_list_header.server_display_name"
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.08)",
          "borderRadius": 60,
          "height": 120,
          "justifyContent": "center",
          "width": 120,
        }
      }
    >
      <Icon
        name="alert-circle-outline"
        style={
          Object {
            "color": "rgba(255,255,255,0.48)",
            "fontSize": 72,
            "lineHeight": 72,
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Metropolis-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 28,
          },
          Object {
            "color": "#ffffff",
            "marginTop": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      Couldn't load Test Team!
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          Object {
            "color": "#ffffff",
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      There was a problem loading content for this team.
    </Text>
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "marginTop": 24,
            },
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 4,
                "flex": 0,
                "justifyContent": "center",
              },
              Object {
                "height": 48,
                "paddingHorizontal": 24,
                "paddingVertical": 14,
              },
              Object {
                "backgroundColor": "#ffffff",
              },
            ],
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignItems": "center",
                "fontFamily": "OpenSans-SemiBold",
                "fontWeight": "600",
                "justifyContent": "center",
                "padding": 1,
                "textAlignVertical": "center",
              },
              Object {
                "fontSize": 16,
                "lineHeight": 16,
                "marginTop": 1,
              },
              Object {
                "color": "#1c58d9",
              },
            ]
          }
        >
          Retry
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`components/channel_list should render team error 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "maxWidth": "100%",
      },
    }
  }
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#1e325c",
      "flex": 1,
      "maxWidth": "100%",
      "paddingLeft": 18,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "marginLeft": 0,
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "marginLeft": 0,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgba(255,255,255,0.64)",
          "fontFamily": "OpenSans-SemiBold",
          "fontSize": 11,
          "fontWeight": "600",
          "lineHeight": 16,
        }
      }
      testID="channel_list_header.server_display_name"
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255,255,255,0.12)",
        "borderRadius": 8,
        "flex": 1,
        "flexDirection": "row",
        "height": 40,
        "justifyContent": "flex-start",
        "marginVertical": 20,
        "maxHeight": 40,
        "padding": 8,
        "width": "100%",
      }
    }
  >
    <Icon
      name="magnify"
      style={
        Object {
          "color": "rgba(255,255,255,0.72)",
          "fontSize": 24,
          "width": 24,
        }
      }
    />
    <TextInput
      placeholder="Find Channels"
      placeholderTextColor="rgba(255,255,255,0.72)"
      style={
        Array [
          Array [
            Object {
              "fontFamily": "OpenSans",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Object {
              "textAlignVertical": "center",
            },
          ],
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "color": "#ffffff",
            "flex": 1,
            "height": 40,
            "marginLeft": 5,
            "marginTop": -2,
          },
        ]
      }
    />
  </View>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <Icon
          name="message-text-outline"
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 24,
              "lineHeight": 28,
            }
          }
        />
        <Text
          style={
            Array [
              Array [
                Object {
                  "fontFamily": "OpenSans-SemiBold",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                },
              ],
              Object {
                "color": "#ffffff",
                "paddingLeft": 12,
              },
            ]
          }
        >
          Threads
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.08)",
          "borderRadius": 60,
          "height": 120,
          "justifyContent": "center",
          "width": 120,
        }
      }
    >
      <Icon
        name="alert-circle-outline"
        style={
          Object {
            "color": "rgba(255,255,255,0.48)",
            "fontSize": 72,
            "lineHeight": 72,
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Metropolis-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 28,
          },
          Object {
            "color": "#ffffff",
            "marginTop": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      Couldn't load categories for 
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          Object {
            "color": "#ffffff",
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      There was a problem loading content for this server.
    </Text>
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "marginTop": 24,
            },
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 4,
                "flex": 0,
                "justifyContent": "center",
              },
              Object {
                "height": 48,
                "paddingHorizontal": 24,
                "paddingVertical": 14,
              },
              Object {
                "backgroundColor": "#ffffff",
              },
            ],
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignItems": "center",
                "fontFamily": "OpenSans-SemiBold",
                "fontWeight": "600",
                "justifyContent": "center",
                "padding": 1,
                "textAlignVertical": "center",
              },
              Object {
                "fontSize": 16,
                "lineHeight": 16,
                "marginTop": 1,
              },
              Object {
                "color": "#1c58d9",
              },
            ]
          }
        >
          Retry
        </Text>
      </View>
    </View>
  </View>
</View>
`;
